src/os_version_check/darwin_impl.rs

```diff
@@ -0,0 +1,590 @@
+#[inline]
+pub(super) fn current_version() -> u32 {
+    // Cache the lookup for performance.
+    //
+    // 0.0.0 is never gonna be a valid version, so we use that as our sentinel value.
+    static CURRENT_VERSION: AtomicU32 = AtomicU32::new(0);
+
+    // We use relaxed atomics, it doesn't matter if multiple threads end up racing to read or write
+    // the version, `lookup_version` should be idempotent and always return the same value.
+    //
+    // `compiler-rt` uses `dispatch_once`, but that's overkill for the reasons above.
+    let version = CURRENT_VERSION.load(Ordering::Relaxed);
+    if version == 0 {
+        let version = lookup_version().get();
+        CURRENT_VERSION.store(version, Ordering::Relaxed);
+        version
+    } else {
+        version
+    }
+}
+
+#[cold]
+// Use `extern "C"` to abort on panic, allowing `current_version` to be free of panic handling.
+pub(super) extern "C" fn lookup_version() -> NonZero<OSVersion> {
+    // Since macOS 10.15, libSystem has provided the undocumented `_availability_version_check` via
+    // `libxpc` (zippered, so requires platform parameter to differentiate between on macOS and Mac
+    // Catalyst) for doing the version lookup, though it's usage may be a bit dangerous, see:
+    // - https://reviews.llvm.org/D150397
+    // - https://github.com/llvm/llvm-project/issues/64227
+    //
+    // So instead, we use the safer approach of reading from `sysctl` (which is faster), and if that
+    // fails, we fall back to the property list (this is what `_availability_version_check` does
+    // internally).
+    let version = version_from_sysctl().unwrap_or_else(version_from_plist);
+
+    // Use `NonZero` to try to make it clearer to the optimizer that this will never return 0.
+    NonZero::new(version).expect("version cannot be 0.0.0")
+}
+
+/// Look up the current OS version(s) from `/System/Library/CoreServices/SystemVersion.plist`.
+///
+/// More specifically, from the `ProductVersion` and `iOSSupportVersion` keys, and from
+/// `$IPHONE_SIMULATOR_ROOT/System/Library/CoreServices/SystemVersion.plist` on the simulator.
+///
+/// This file was introduced in macOS 10.3, which is well below the minimum supported version by
+/// `rustc`, which is currently macOS 10.12.
+///
+/// # Panics
+///
+/// Panics if reading or parsing the version fails (or if the system was out of memory).
+///
+/// We deliberately choose to panic, as having this silently return an invalid OS version would be
+/// impossible for a user to debug.
+#[allow(non_upper_case_globals, non_snake_case)]
+pub(super) fn version_from_plist() -> OSVersion {
+    #[allow(clippy::upper_case_acronyms)]
+    enum FILE {}
+
+    const SEEK_END: c_int = 2;
+
+    const RTLD_LAZY: c_int = 0x1;
+    const RTLD_LOCAL: c_int = 0x4;
+
+    // SAFETY: Same signatures as in `libc`.
+    //
+    // NOTE: We do not need to link these; that will be done by `std` by linking `libSystem`
+    // (which is required on macOS/Darwin).
```

Comment:
Could you double check that building a `#![no_std]` crate with this version of `compiler_builtins` links without errors? We won't have the system libraries in that case so it seems like this would fail. There aren't any other cases where we call `libc` functions in this crate.

Reply:
Ah, Bjorn just said the same thing.